homeassistant/components/sensor/__init__.py

```diff
@@ -273,6 +277,67 @@ def state(self) -> Any:
         """Return the state of the sensor and perform unit conversions, if needed."""
         unit_of_measurement = self.native_unit_of_measurement
         value = self.native_value
+        device_class = self.device_class
+
+        # We have an old non-datetime value, warn about it and convert it during
+        # the deprecation period.
+        if (
+            value is not None
+            and device_class in (DEVICE_CLASS_DATE, DEVICE_CLASS_TIMESTAMP)
+            and not isinstance(value, (date, datetime))
+        ):
+            # Deprecation warning for date/timestamp device classes
+            if not self.__datetime_as_string_deprecation_logged:
+                caller = inspect.stack()[2]
+                module = inspect.getmodule(caller[0])
+                if module and "custom_components" in module.__file__:
+                    report_issue = "report it to the custom component author."
+                else:
+                    report_issue = (
+                        "create a bug report at "
+                        "https://github.com/home-assistant/core/issues?q=is%3Aopen+is%3Aissue"
+                    )
+                _LOGGER.warning(
+                    "%s is providing a string for its state, while the device "
+                    "class is '%s', this is not valid and will be unsupported "
+                    "from Home Assistant 2022.2. Please %s",
+                    module.__name__ if module else "Unknown",
+                    device_class,
+                    report_issue,
+                )
+                self.__datetime_as_string_deprecation_logged = True
+
+            # Anyways, lets validate the date at least..
+            try:
+                value = ciso8601.parse_datetime(str(value))
+            except (ValueError, IndexError) as error:
+                raise ValueError(
+                    f"Invalid date/datetime: {self.entity_id} provide state '{value}', "
+                    f"while it has device class '{device_class}'"
+                ) from error
+
+            # Convert the date object to a standardized state string.
+            if device_class == DEVICE_CLASS_DATE:
+                return value.date().isoformat()
+            return value.isoformat(timespec="seconds")
+
+        # Received a datetime
+        if value is not None and isinstance(value, datetime):
+            if device_class != DEVICE_CLASS_TIMESTAMP:
+                raise ValueError(
+                    f"Invalid datetime: {self.entity_id} provides a {type(value)}"
+                    "state, however, does not have a timestamp device class"
+                )
+            return value.isoformat(timespec="seconds")
```

Comment:
I'd drop the `isinstance` check and use `try... except` to catch `AttributeError` and `TypeError` and raise the `ValueError` with the message if there's an error. We only require datetime for timestamp device class. The inverse relationship is not a requirement. Ie, we don't require all datetimes to have timestamp device class.
```suggestion
        if value is not None and device_class == DEVICE_CLASS_TIMESTAMP:
            try:
                return value.isoformat(timespec="seconds")
            except (AttributeError, TypeError) as err:
                raise ValueError(
                    f"Invalid datetime: {self.entity_id} has a timestamp device class"
                    f"but does not provide a datetime state but {type(value)}"
                ) from err
```

Reply:
> inverse relationship is not a requirement. Ie, we don't require all datetimes to have timestamp device class.

That is not true? `datetime` and `date` objects are not valid native states otherswise

Reply:
They are converted to string in the base entity class.

Reply:
That is true. Ok, let's do this, for now, this is already strict enough indeed.

Reply:
I've made the same adjustment for the date variant.

The downside of this approach is that explicit type checking is removed, this I had to add type ignores to make mypy shut up.